Xenojelturret&larvajelly

Kamatii opened the pull request:
This is my first time coding, and making a pr, sorry if there are mistakes, talked with spaghetti beforehand.

Adds a new turret for xenos, "xeno_turret_jel". This turret shoots targets with a xeno spit, upon impact it injects them with larvajelly, a reagent that increases the growth speed of the larva inside host. Targetting looks for 2 conditions, target must be alive, must be infected with larvae, must not contain larvajelly in their system. The turret has a recoloured acid turret sprite.
Removed the damage taken from larvajelly.
Why It's Good For The Game
Code bounty by the host.

Changelog
🆑 
add: Added xeno jelly turret
add: Added xeno jelly turret build ability
add: Added xeno jelly turret ammo (spit)
del: Removed larvajelly's and it's prepared varient's damage to humans
add: Added larvajelly, and larvajellyprepared's new color and ability
imageadd: Added xeno jelly turret sprites
/🆑

Kamatii committed:
Jel turret code

Kamatii committed:
adding it's new ammo

Kamatii committed:
Define jel turret ammo color

Kamatii committed:
Build jel turret ability

Kamatii committed:
who can build it

Kamatii committed:
larva jelly and larva jelly prepared are nolonger harmful, but lewd

Kamatii committed:
Fixed color and removed color datum

Kamatii committed:
adding sprites

Kamatii committed:
You can taste slimy, to, You can taste slime

ProbablySpaghetti commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1205,6 +1205,79 @@
+	/// Pyschic point cost
+	var/psych_cost = XENO_TURRET_PRICE
+
+/datum/action/xeno_action/activable/build_turret_jel/can_use_ability(atom/A, silent, override_flags)
```
I'd suggest making this a subtype of the above action, `/datum/action/xeno_action/activable/build_turret`. That way you don't actually need to redefine this whole `can_use_ability()` proc and all of the variables, you can just override `use_ability()`.

ProbablySpaghetti commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1205,6 +1205,79 @@
+//////////////////////////
+
+/datum/action/xeno_action/activable/build_turret_jel
+	name = "Secrete jel turret"
```
What's Jel?
(Larva jelly presumably, but someone playing ingame might not know that.)

ProbablySpaghetti commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1931,6 +1931,41 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
+/datum/ammo/xeno/larvae_jelly/proc/set_reagents()
+	spit_reagents = list(/datum/reagent/consumable/larvajelly = reagent_transfer_amount)
+
+/datum/ammo/xeno/larvae_jelly/on_hit_mob(mob/living/victim, obj/projectile/proj)
+
+	if(!istype(victim) || victim.stat == DEAD || victim.issamexenohive(proj.firer))
```
Can delete this empty line here
```suggestion
/datum/ammo/xeno/larvae_jelly/on_hit_mob(mob/living/victim, obj/projectile/proj)
	if(!istype(victim) || victim.stat == DEAD || victim.issamexenohive(proj.firer))
```

ProbablySpaghetti commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1931,6 +1931,41 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
+/datum/ammo/xeno/larvae_jelly
+	name = "blob of larvae jelly"
+	icon_state = "larvajelly"
+	ping = "ping_x"
+	flags_ammo_behavior = AMMO_XENO|AMMO_SKIPS_ALIENS|AMMO_EXPLOSIVE
+	var/danger_message = "<span class='danger'>A blob of larvae jelly lands with a splat!</span>"
+	armor_type = "bio"
```
`ping` and `armor_type` are the same as on `/datum/ammo/xeno`, so they can be removed from here.

ProbablySpaghetti commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1931,6 +1931,41 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
+	reagent_transfer_amount = 5
+
+///Set up the list of reagents the spit transfers upon impact
+/datum/ammo/xeno/larvae_jelly/proc/set_reagents()
```
The fact that this proc is redefined three separate times rather than just being on `/datum/ammo/xeno` is... awful.
(Not your fault, just wanted to mention it.)

ProbablySpaghetti commented on code/modules/reagents/reagents/food.dm:
```diff
@@ -406,28 +406,36 @@
-	nutriment_factor = 0
-	color = "#66801e"
-	taste_description = "burning"
+	nutriment_factor = 1
```
Same as above, `nutriment_factor` is also `1` on `/datum/reagent/consumable`, so it can be removed from here.

ProbablySpaghetti commented on code/modules/reagents/reagents/food.dm:
```diff
@@ -406,28 +406,36 @@
 /datum/reagent/consumable/larvajelly
 	name = "Larva Jelly"
-	description = "The blood and guts of a xenomorph larva blended into a paste. Drinking this is bad for you."
+	description = "A hormonal, sticky, slick jelly made inside xeno's reproductory organs, containing nutrients and pheromones to make the host's body nurish and nurture the larva inside."
```
```suggestion
	description = "A hormonal, sticky, slick jelly made inside a xeno's reproductory organs, containing nutrients and pheromones to make the host's body nourish and nurture the larva inside."
```

ProbablySpaghetti commented on code/modules/reagents/reagents/food.dm:
```diff
@@ -406,28 +406,36 @@
-	name = "Prepared Larva Jelly"
-	description = "A delicious blend of xenomorphic entrails and acid, denatured by exposure to high-frequency radiation. Probably has some uses."
+	name = "Condensed Larva Jelly"
+	description = "A more potent, less slimy, but much more slippery version of Larva Jelly. The hormones are lost after cooking, leaving only the pheromones. This slipper, sticky liquid probably has some uses."
```
Probably don't need to restate this part
```suggestion
	description = "A more potent, less slimy, but much more slippery version of Larva Jelly. The hormones are lost after cooking, leaving only the pheromones. Probably has some uses."
```

ProbablySpaghetti commented on code/modules/xenomorph/xeno_structures.dm:
```diff
@@ -452,3 +452,217 @@
+///// xeno jell turret
+//////////////////////
+
+/obj/structure/xeno/resin/xeno_turret_jel
```
Same thing here as with `/datum/action/xeno_action/activable/build_turret_jel`. By not making it a subtype you're having to redefine a *lot* of code.
If this was `/obj/structure/xeno/resin/xeno_turret/jel` or something you'd only need to change a few things for it to work.

ProbablySpaghetti commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1931,6 +1931,41 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
 
+/datum/ammo/xeno/larvae_jelly
+	name = "blob of larvae jelly"
+	icon_state = "larvajelly"
```
Probably don't need a whole new icon state for this, since it's just the same one as `sticky`.
```suggestion
	icon_state = "sticky"
```

Kamatii committed:
Reverting the icon for projectile since it uses the same one as "sticky"

Kamatii committed:
ammi icon fix

Kamatii committed:
set jelly turret as child, squeezed and named it jelly instead of jel

Kamatii committed:
fixed larvajelly and larvajellyprepared

Kamatii committed:
ammo fix and trim

Kamatii committed:
Squeezing the building code, fixing cast ability, renaming ammo from larvae_jelly to larva_jelly

Kamatii committed:
fix desc and larvajelly-larvajellyprepared

Kamatii committed:
I can't squeeze get_target(), it sticks to one single target and won't let go

it works fine, if we have to squeeze it down, someone please do, I really don't know how, otherwise, with this code it works without issue as I checked.

ProbablySpaghetti commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1195,15 +1195,38 @@
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!T.check_alien_construction(X, planned_building = /obj/structure/xeno/resin/xeno_turret/jelly) || !T.check_disallow_alien_fortification(X))
+		return FALSE
```
The `check_disallow_alien_fortification()` can be removed here, since it's already checked just above on the parent:
https://github.com/ProbablySpaghetti/XRF-13/blob/8d2f5f1106b0986f83e628ec3940cf91914bc4ba/code/modules/mob/living/carbon/xenomorph/abilities.dm#L1179-L1180
```suggestion
	if(!T.check_alien_construction(X, planned_building = /obj/structure/xeno/resin/xeno_turret/jelly))
		return FALSE
```

ProbablySpaghetti commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1195,15 +1195,38 @@
+	succeed_activate()
 
+/datum/action/xeno_action/activable/build_turret/proc/building_turret(atom/A, mob/living/carbon/xenomorph/X)
 	to_chat(owner, "<span class='xenowarning'>We build a new acid turret, spending 100 psychic points in the process</span>")
```
This should probably be changed to reference `plasma_cost`, that way it's always consistent. (Same on line 1226 below)
```suggestion
	to_chat(owner, "<span class='xenowarning'>We build a new acid turret, spending [plasma_cost] psychic points in the process</span>")
```

ProbablySpaghetti commented on code/modules/projectiles/ammo_datums.dm:
```diff
@@ -1544,12 +1544,10 @@ GLOBAL_LIST_INIT(no_sticky_resin, typecacheof(list(/obj/item/clothing/mask/faceh
 */
 /datum/ammo/xeno
 	icon_state = "neurotoxin"
-	ping = "ping_x"
 	damage_type = TOX
 	flags_ammo_behavior = AMMO_XENO
 	var/added_spit_delay = 0 //used to make cooldown of the different spits vary.
 	var/spit_cost = 5
-	armor_type = "bio"
```
Add these two back, but remove them from `/datum/ammo/xeno/larva_jelly`, since it inherits.

ProbablySpaghetti commented on code/modules/xenomorph/xeno_structures.dm:
```diff
@@ -452,3 +450,90 @@
+	firerate = 20
+
+/obj/structure/xeno/resin/xeno_turret/jelly/Initialize(mapload, hivenumber = XENO_HIVE_NORMAL)
+	.=..()
```
Very nitpicky, but could you add some spaces here and on line 478 so that it's easier to read?
```suggestion
	. = ..()
```

ProbablySpaghetti commented on code/modules/xenomorph/xeno_structures.dm:
```diff
@@ -452,3 +450,90 @@
+			distance = buffer_distance
+			. = nearby_hostile
+
+///Return TRUE if a possible target is near
+/obj/structure/xeno/resin/xeno_turret/jelly/proc/scanjel()
+	potential_hostiles.Cut()
+	for (var/mob/living/carbon/human/nearby_human AS in cheap_get_humans_near(src, TURRET_SCAN_RANGE))
+		if(nearby_human.stat == DEAD || nearby_human.reagents.has_reagent(/datum/reagent/consumable/larvajelly) || !locate(/obj/item/alien_embryo) in nearby_human)
+			continue
+		potential_hostiles += nearby_human
```
You should just be able to override `scan()` with this code, rather than having a new proc for it.
```suggestion
///Return TRUE if a possible target is near
/obj/structure/xeno/resin/xeno_turret/jelly/scan()
	potential_hostiles.Cut()
	for (var/mob/living/carbon/human/nearby_human AS in cheap_get_humans_near(src, TURRET_SCAN_RANGE))
		if(nearby_human.stat == DEAD || nearby_human.reagents.has_reagent(/datum/reagent/consumable/larvajelly) || !locate(/obj/item/alien_embryo) in nearby_human)
			continue
		potential_hostiles += nearby_human
```

ProbablySpaghetti commented on code/modules/xenomorph/xeno_structures.dm:
```diff
@@ -452,3 +450,90 @@
+	ammo = GLOB.ammo_list[/datum/ammo/xeno/larva_jelly]
+	set_light(2, 2, LIGHT_COLOR_PURPLE)
+
+/obj/structure/xeno/resin/xeno_turret/jelly/process()
+	//Turrets regen some HP, every 2 sec
+	.=..()
+	if(world.time > last_scan_time + TURRET_SCAN_FREQUENCY)
+		scanjel()
+		last_scan_time = world.time
+	if(!hostile || hostile.reagents.has_reagent(/datum/reagent/consumable/larvajelly) || !locate(/obj/item/alien_embryo) in hostile)
+		if(last_hostile)
+			set_last_hostile(null)
+		return
```
If you *do* override it, then you can remove this cunk of the `process()` code here:
```suggestion
/obj/structure/xeno/resin/xeno_turret/jelly/process()
	//Turrets regen some HP, every 2 sec
	. = ..()
	if(!hostile || hostile.reagents.has_reagent(/datum/reagent/consumable/larvajelly) || !locate(/obj/item/alien_embryo) in hostile)
		if(last_hostile)
			set_last_hostile(null)
		return
```

ProbablySpaghetti replied:
Can also remove this `return`, since `return` is always implied at the end of a proc. (It'll return no matter what happens.)
```suggestion
		if(last_hostile)
			set_last_hostile(null)
```

ProbablySpaghetti commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1195,15 +1195,38 @@
+	mechanics_text = "Creates a new xeno larva jelly turret for 150 points"
+	ability_name = "secrete larva jelly turret"
+
+/datum/action/xeno_action/activable/build_turret/jelly/can_use_ability(atom/A, silent, override_flags, mob/living/carbon/xenomorph/X, turf/T)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!T.check_alien_construction(X, planned_building = /obj/structure/xeno/resin/xeno_turret/jelly) || !T.check_disallow_alien_fortification(X))
```
Need to define `T` here since it doesn't carry over from the parent.
```suggestion
/datum/action/xeno_action/activable/build_turret/jelly/can_use_ability(atom/A, silent, override_flags, mob/living/carbon/xenomorph/X, turf/T)
	. = ..()
	if(!.)
		return FALSE
	var/turf/T = get_turf(A)
	if(!T.check_alien_construction(X, planned_building = /obj/structure/xeno/resin/xeno_turret/jelly) || !T.check_disallow_alien_fortification(X))
```

ProbablySpaghetti commented on code/modules/xenomorph/xeno_structures.dm:
```diff
@@ -452,3 +450,90 @@
+
+/obj/structure/xeno/resin/xeno_turret/jelly/turret_message()
+	if(!TIMER_COOLDOWN_CHECK(src, COOLDOWN_XENO_TURRETS_ALERT))
+		associated_hive.xeno_message("Our [name] is injecting a nearby impregnated host [hostile] at [get_area(hostile)] (X: [hostile.x], Y: [hostile.y]).", "xenoannounce", 5, FALSE, hostile, 'sound/voice/alien_help1.ogg', FALSE, null, /obj/screen/arrow/turret_attacking_arrow)
```
I'd suggest removing the sound from this, since by design it's gonna be firing a *lot*, and having a xeno shrieking in your ear every isn't that enjoyable. :P
```suggestion
		associated_hive.xeno_message("Our [name] is injecting a nearby impregnated host [hostile] at [get_area(hostile)] (X: [hostile.x], Y: [hostile.y]).", "xenoannounce", 5, FALSE, hostile, null, FALSE, null, /obj/screen/arrow/turret_attacking_arrow)
```

ProbablySpaghetti reviewed (commented):
Mostly nitpicks this time, should be good to testmerge once these are cleared up.

Kamatii committed:
Removing notification sound

Kamatii committed:
removing notification sound 2

Kamatii committed:
adding var T to build turret

Kamatii committed:
remove "Return" from jelly/process

Kamatii committed:
changing scanjel to scan

Kamatii committed:
putting space for .=..() for better reading

Kamatii committed:
oh my bad x3, I removed the wrong lines, fixed now

Kamatii committed:
removed extra code

Kamatii committed:
fixing T again

Kamatii committed:
[plasma_cost] instead of text
